advanced_topics.qmd

```diff
@@ -266,15 +266,59 @@ pums_person |>
   collect()
 ```
 
+Another example is where you've created a model based on a subset of your data and you want to use it to make predictions.
+Let's say we've created a linear model, using age and employment status to predict commute time for respondents in California in 2005.
+
+```{r}
+#| label: predict-commute-time
+model_data <- open_dataset("data/person/") |>
+  filter(year==2005, location=="ca") |>
+  transmute(
+    commute_time = JWMNP,
+    age = AGEP,
+    self_employed = stringr::str_detect(as.character(COW), "Self-employed")
+  ) |>
+  collect()
+
+my_model <- lm(commute_time ~ age + self_employed, data = model_data)
+summary(my_model)
+```
+
+Let's say we want to use this model to predict commute times in California in 2021.
+We can create a UDF which calls `predict()` to add the predicted commute time value to the data.
+
+```{r}
+#| label: predict-commute-time
+register_scalar_function(
+  "predict_commute_time",
+  function(context, age, self_employed) {
+    predict(my_model, newdata = data.frame(age, self_employed))
+  },
+  in_type = schema(age = int32(), self_employed = boolean()),
+  out_type = float64(),
+  auto_convert = TRUE
+)
+
+open_dataset("data/person/") |>
+  filter(year==2021, location=="ca") |>
+  transmute(
+    age = AGEP,
+    self_employed = stringr::str_detect(as.character(COW), "Self-employed"),
+    commute_time_predicted = predict_commute_time(age, self_employed)
+  ) |>
+  collect() |>
+  head()
+```
+
```

Comment:
Should we add some discussion? Gratuitous plot of predicted/fitted vs. actual values?

Reply:
I added it but the plot looks a bit shit as the model is shit :joy:
Not sure if it's worth putting in the effort to make the model better to produce a better plot.
Let me know what you think!

Reply:
I wonder if it would be worth 15 minutes of effort to try to make a reasonable model. Suppose that: calling back to the example in the Getting Started chapter, we fit a model where the outcome is whether someone works from home. We fit the model with 2015 data or something like we do here, and then show that it reasonably well predicts the outcome in 2019. Then we show with 2020 and how it doesn't. 

I don't think this is worth hours of effort, but it would be a nice story if it weren't hard to do. (And think of the page count from adding 3 new plots!)

Thoughts?

Reply:
It'll take _me_ more than 15 minutes as I'll end up agonising over exactly how good the model needs to be, lol, but if you take this one and I take the "produce the PDF in greyscale" instead, that works for me?